transmart-core-db/grails-app/services/org/transmartproject/db/tree/TreeCacheService.groovy

```diff
@@ -131,30 +130,35 @@ class TreeCacheService {
         criteria.addOrder(Order.desc('level'))
         criteria.addOrder(Order.desc('fullName'))
 
-        def t1 = new Date()
+        def stopWatch = new StopWatch('Collecting tree nodes.')
+        stopWatch.start('Fetching nodes from the database.')
         List<I2b2Secure> i2b2Nodes = criteria.getExecutableCriteria(sessionFactory.currentSession).list()
-        def t2 = new Date()
-        log.info "Found ${i2b2Nodes.size()} nodes. Query took ${t2.time - t1.time} ms."
-
-        List<TreeNode> forest = buildForest(i2b2Nodes)
-        def t3 = new Date()
-        log.debug "Forest growing took ${t3.time - t2.time} ms."
+        stopWatch.stop()
+        stopWatch.start('Filtering nodes user has access to.')
+        List<I2b2Secure> accessibleI2b2Nodes = i2b2Nodes.findAll { I2b2Secure i2b2Secure ->
```

Comment:
Logically fetching all and filtering on the onces you have access to is the same as fetching the once you have access to, but I am really in favour of have both functionalities:
- checking for an entity, if the user has sufficient permissions for it;
- fetching all entities (of a certain type/matching certain constraints) that the user has sufficient permissions for.

Reply:
The first one is currently available in the `AccessControlChecks`/`User` class, the second one seems appropriate for specific resources, such as this (for tree nodes).